Ability to administrate the elements from menus

alecslupu opened the pull request:
#### :tophat: What? Why?
Often the client need is to hide / reorder menu elements. This PR will add support for doing the exact thing. 
Each menu within the application has a name. Each Menu entry has its own identifier, which later can be used as selector. 

The following snippet will add 2 different menu entries, one with position 1, and second with position 2
```
Decidim.menu :admin_participatory_process_attachments_menu do |menu|
  menu.add_item :participatory_process_attachment_collections,
                I18n.t("attachment_collections", scope: "decidim.admin.menu.participatory_processes_submenu"),
                decidim_admin_participatory_processes.participatory_process_attachment_collections_path(current_participatory_space),
                active: is_active_link?(decidim_admin_participatory_processes.participatory_process_attachment_collections_path(current_participatory_space)),
                if: allowed_to?(:read, :attachment_collection)

  menu.add_item :participatory_process_attachments,
                I18n.t("attachment_files", scope: "decidim.admin.menu.participatory_processes_submenu"),
                decidim_admin_participatory_processes.participatory_process_attachments_path(current_participatory_space),
                active: is_active_link?(decidim_admin_participatory_processes.participatory_process_attachments_path(current_participatory_space)),
                if: allowed_to?(:read, :attachment)
end 
```

The following snippet, will prevent rendering the menu entry with identifier `user_groups` from the `admin_user_menu`. 
```
Decidim.menu :admin_user_menu do |menu|
  menu.remove_item :user_groups
end
```

The following snippet, will move the specified element entry before / after the specified anchor: 

```
Decidim.menu :admin_user_menu do |menu|
  menu.move_after after: :users, element: :moderated_users
  menu.move_before element: :authorization_workflows, before: :officializations
end

```

#### :pushpin: Related Issues
*Link your PR to an issue*
- #7368
- #7382
- https://meta.decidim.org/processes/roadmap/f/122/proposals/15774
- https://meta.decidim.org/processes/roadmap/f/122/proposals/13246

#### Testing
1. Visit the admin interface 
2. In the decidim initializer of a rails application, any kind of configuration like the examples above
3. Restart the webserver 
4. Visit the page that renders that menu
5. Observe the changes 

#### :clipboard: Checklist
:rotating_light: Please review the [guidelines for contributing](../CONTRIBUTING.adoc) to this repository.

- [ ] :question: **CONSIDER** adding a unit test if your PR resolves an issue.
- [ ] :heavy_check_mark: **DO** check open PR's to avoid duplicates.
- [ ] :heavy_check_mark: **DO** keep pull requests small so they can be easily reviewed.
- [ ] :heavy_check_mark: **DO** build locally before pushing.
- [ ] :heavy_check_mark: **DO** make sure tests pass.
- [x] :heavy_check_mark: **DO** make sure any new changes are documented in `docs/`.
- [ ] :heavy_check_mark: **DO** add and modify seeds if necessary.
- [x] :heavy_check_mark: **DO** add CHANGELOG upgrade notes if required.
- [ ] :heavy_check_mark: **DO** add to GraphQL API if there are new public fields.
- [x] :heavy_check_mark: **DO** add link to MetaDecidim if it's a new feature.
- [ ] :x:**AVOID** breaking the continuous integration build.
- [ ] :x:**AVOID** making significant changes to the overall architecture.

### :camera: Screenshots

Original menu:
![original_user_menu](https://user-images.githubusercontent.com/105683/110542602-e9c03300-8131-11eb-8926-0f0238da8fe9.png)

Modified menu using snippet:
```ruby 
Decidim.menu :user_menu do |menu|
  menu.remove_item :user_interests
  menu.move :own_user_groups, after: :account
  menu.move :authorizations, before: :notifications_settings
end
```
![modified_user_menu](https://user-images.githubusercontent.com/105683/110542576-e036cb00-8131-11eb-9c01-136b877275a9.png)

### Additional information 
Decidim will define 33 menus (and submenus) on the entire application : 
```
[:admin_assemblies_components_menu, :admin_assemblies_menu, :admin_conferences_components_menu, :admin_consultation_components_menu, :admin_consultation_menu, :admin_consultation_question_menu, :admin_global_moderation_menu, :admin_initiatives_components_menu, :admin_initiatives_menu, :admin_menu, :admin_participatory_process_attachments_menu, :admin_participatory_process_components_menu, :admin_participatory_process_group_menu, :admin_participatory_process_menu, :admin_participatory_processes_menu, :admin_questions_menu, :admin_settings_menu, :admin_template_types_menu, :admin_user_menu, :admin_votings_components_menu, :assemblies_admin_attachments_menu, :assemblies_admin_menu, :conferences_admin_attachments_menu, :conferences_admin_menu, :conferences_admin_registrations_menu, :decidim_initiative_menu, :decidim_voting_menu, :decidim_votings_attachments_menu, :impersonate_menu, :menu, :system_menu, :user_menu, :workflows_menu]
```
The list of menus and identifiers: 

```
Decidim.menu :admin_menu do |menu|
  menu.add_item :dashboard
  menu.add_item :participatory_processes
  menu.add_item :assemblies
  menu.add_item :conferences
  menu.add_item :consultations
  menu.add_item :votings  
  menu.add_item :initiatives
  menu.add_item :moderations
  menu.add_item :static_pages
  menu.add_item :impersonatable_users
  menu.add_item :newsletters
  menu.add_item :edit_organization
  menu.add_item :questionnaire_templates
  menu.add_item :logs
end

Decidim.menu :menu do |menu|
  menu.add_item :root
  menu.add_item :pages
  menu.add_item :participatory_processes
  menu.add_item :assemblies
  menu.add_item :conferences
  menu.add_item :consultations
  menu.add_item :initiatives
  menu.add_item :votings
end

Decidim.menu :admin_settings_menu do |menu|
  menu.add_item :edit_organization
  menu.add_item :edit_organization_appearance
  menu.add_item :edit_organization_homepage
  menu.add_item :scopes
  menu.add_item :scope_types
  menu.add_item :areas
  menu.add_item :area_types
  menu.add_item :help_sections
end

Decidim.menu :admin_user_menu do |menu|
  menu.add_item :users
  menu.add_item :user_groups
  menu.add_item :officializations
  menu.add_item :impersonatable_users, submenu: :impersonate_menu
  menu.add_item :moderated_users
  menu.add_item :authorization_workflows, submenu: :workflows_menu
end

Decidim.menu :impersonate_menu do |menu|
  menu.add_item :conflicts
end

Decidim.menu :workflows_menu do |menu|
  # dynamically generated by activated workflows
  menu.add_item :id_documents
  menu.add_item :postal_letter
  menu.add_item :csv_census
  menu.add_item :sms
end

Decidim.menu :admin_global_moderation_menu do |menu|
  menu.add_item :moderations
  menu.add_item :hidden_moderations
end
Decidim.menu :assemblies_admin_attachments_menu do |menu|
  menu.add_item :assembly_attachment_collections
  menu.add_item :assembly_attachments
end
Decidim.menu :admin_assemblies_components_menu do |menu|
  # dynamically generated based on installed component's name and id of the record
  menu.add_item "meetings_123"
  ....
end
Decidim.menu :admin_conferences_components_menu do |menu|
  # dynamically generated based on installed component's name and id of the record
  menu.add_item "meetings_123"
  ....
end
Decidim.menu :assemblies_admin_menu do |menu|
  menu.add_item :edit_assembly
  menu.add_item :components, submenu: :admin_assemblies_components_menu
  menu.add_item :categories
  menu.add_item :attachments, submenu: :assemblies_admin_attachments_menu 
  menu.add_item :assembly_members
  menu.add_item :assembly_user_roles
  menu.add_item :participatory_space_private_users
  menu.add_item :moderations
end
Decidim.menu :admin_assemblies_menu do |menu|
  menu.add_item :assemblies
  menu.add_item :assemblies_types
  menu.add_item :edit_assemblies_settings
end
Decidim.menu :conferences_admin_registrations_menu do |menu|
  menu.add_item :conference_registration_types
  menu.add_item :conference_conference_registrations
  menu.add_item :conference_conference_invites
  menu.add_item :edit_conference_diploma
end
Decidim.menu :conferences_admin_attachments_menu do |menu|
  menu.add_item :conference_attachment_collections
  menu.add_item :conference_attachments
end      
Decidim.menu :conferences_admin_menu do |menu|
  menu.add_item :edit_conference
  menu.add_item :components, submenu: :admin_conferences_components_menu
  menu.add_item :categories
  menu.add_item :attachments, submenu: :conferences_admin_attachments_menu
  menu.add_item :conference_media_links
  menu.add_item :conference_partners
  menu.add_item :conference_speakers
  menu.add_item :registrations, submenu: :conferences_admin_registrations_menu 
  menu.add_item :conference_user_roles
  menu.add_item :moderations
end
Decidim.menu :admin_consultation_menu do |menu|
  menu.add_item :edit_consultation
  menu.add_item :consultation_questions
  menu.add_item :results_consultation
end
Decidim.menu :admin_consultation_question_menu do |menu|
  menu.add_item :edit_consultation
  menu.add_item :consultation_questions, submenu: :admin_questions_menu 
  menu.add_item :results_consultation
end
Decidim.menu :admin_questions_menu do |menu|
  menu.add_item :edit_question
  menu.add_item :edit_question_configuration
  menu.add_item :responses
  menu.add_item :components, submenu: :admin_consultation_components_menu
  menu.add_item :categories
  menu.add_item :question_attachments
end

Decidim.menu :admin_consultation_components_menu do |menu|
  # dynamically generated based on installed component's name and id of the record
  menu.add_item "meetings_123"
  ....        
end

Decidim.menu :admin_votings_components_menu do |menu|
  # dynamically generated based on installed component's name and id of the record
  menu.add_item "meetings_123"
  ....   
end

Decidim.menu :user_menu do |menu|
  menu.add_item :account
  menu.add_item :notifications_settings
  menu.add_item :authorizations
  menu.add_item :own_user_groups
  menu.add_item :user_interests
  menu.add_item :data_portability
  menu.add_item :delete_account
  menu.add_item :decidim_elections_trustee_zone
  menu.add_item :decidim_votings_polling_officer_zone
end
Decidim.menu :admin_participatory_process_menu do |menu|
  menu.add_item :trustees
end  
Decidim.menu :decidim_votings_attachments_menu do |menu|
  menu.add_item :voting_attachment_collections
  menu.add_item :voting_attachments
end
Decidim.menu :decidim_voting_menu do |menu|
  menu.add_item :edit_voting
  menu.add_item :edit_voting_landing_page
  menu.add_item :components, submenu: :admin_votings_components_menu
  menu.add_item :attachments, submenu: :decidim_votings_attachments_menu
  menu.add_item :voting_polling_stations
  menu.add_item :voting_polling_officers
  menu.add_item :voting_monitoring_committee_members
end    

Decidim.menu :admin_initiatives_components_menu do |menu|
  # dynamically generated based on installed component's name and id of the record
  menu.add_item "meetings_123"
  ....   
end  

Decidim.menu :decidim_initiative_menu do |menu|
  menu.add_item :edit_initiative
  menu.add_item :initiative_committee_requests
  menu.add_item :components, submenu: :admin_initiatives_components_menu
  menu.add_item :initiative_attachments
  menu.add_item :moderations
end
Decidim.menu :admin_initiatives_menu do |menu|
  menu.add_item :initiatives 
  menu.add_item :initiatives_types
end

Decidim.menu :admin_participatory_processes_menu do |menu|
  menu.add_item :participatory_processes
  menu.add_item :participatory_process_groups
end
Decidim.menu :admin_participatory_process_attachments_menu do |menu|
  menu.add_item :participatory_process_attachment_collections 
  menu.add_item :participatory_process_attachments
end
Decidim.menu :admin_participatory_process_components_menu do |menu|
  # dynamically generated based on installed component's name and id of the record
  menu.add_item "meetings_123"
  ....   
end  
Decidim.menu :admin_participatory_process_menu do |menu|
  menu.add_item :edit_participatory_process
  menu.add_item :participatory_process_steps
  menu.add_item :components, submenu: :admin_participatory_process_components_menu
  menu.add_item :categories
  menu.add_item :attachments, submenu: :admin_participatory_process_attachments_menu
  menu.add_item :participatory_process_user_roles
  menu.add_item :participatory_space_private_users
  menu.add_item :moderations
end
Decidim.menu :admin_participatory_process_group_menu do |menu|
  menu.add_item :edit_participatory_process_group
  menu.add_item :edit_participatory_process_group_landing_page
end
Decidim.menu :system_menu do |menu|
  menu.add_item :root
  menu.add_item :organizations
  menu.add_item :admins
  menu.add_item :oauth_applications
end

Decidim.menu :admin_template_types_menu do |menu|
  # dynamically generated based the template_types
end
```

:hearts: Thank you!

alecslupu commented:
@mrcasals, @oriolgual  this now can be reviewed. Thanks

mrcasals commented:
Hi @alecslupu!

I checked the proposals in Metadecidim and they don't seem to be approved by @decidim/product, could you make sure they're approved before reviewing the code?

As per the implementation, I haven't checked it much but the `move` API feels a little verbose:

```ruby
Decidim.menu :admin_user_menu do |menu|
  menu.move_after after: :users, element: :moderated_users
  menu.move_before element: :authorization_workflows, before: :officializations
end
```

Could we have something like this?

```ruby
Decidim.menu :admin_user_menu do |menu|
  menu.move :moderated_users, after: :users
  menu.move :authorization_workflows, before: :officializations
end
```

Internally, the `move` method could be defined something like this:

```ruby
class UncertainPosition < StandardError; end

def move(element, before: nil, after: nil)
  raise UncertainPosition if before && after

  ... # rest of the code
end
```

By raising the exception we'd make sure the app can't boot. 

What do you think?

alecslupu commented:
> ```ruby
> raise UncertainPosition if before && after
> ```

PierreMesure commented:
This is amazing work, @alecslupu! Definitely needs to be reviewed by Product but since the changes in this PR are not usable by the end user (admin in that case), I think they should be considered as a refactor and not subject to Xabier's concerns.

alecslupu commented:
I have asked @carolromero to have a look from their end, and also i am looking to implement the changes asked by @mrcasals , as they make sense :) 

Once this is megred, some comunity gems will not work though ( thinking of term_customizer and navbar )

oriolgual commented:
Could you also add docs in the `/docs` folder about how to add/remove or customize the menus please?

andreslucena commented:
I've reviewed this with @carolromero and we approve this change as @decidim/product 👍🏽 

Just a quick note, I think we should take in account that this is a breaking change as it's changing the API (from `menu.item` to `menu.add_item`). Some alternatives to discuss: 

1) Add this change to the CHANGELOG so implementers/developers that have modules that use this contracts are aware of this change
or
2) Change the `menu.add_item` to `menu.item action: :add` and  `menu.remove_item` to `menu.item action: :remove`  where `:add` is the default param (current behaviour)

mrcasals commented on decidim-core/lib/decidim/menu.rb:
```diff
@@ -42,18 +44,74 @@ def build_for(context)
     #   menu.item "Gestor de Procesos", "/processes", if: admin?
     #
     def item(label, url, options = {})
+      ActiveSupport::Deprecation.warn("Using menu.item in #{@name} context is deprecated. User menu.add_item")
```
```suggestion
      ActiveSupport::Deprecation.warn("Using menu.item in #{@name} context is deprecated. Use menu.add_item")
```

mrcasals commented on decidim-core/lib/decidim/menu.rb:
```diff
@@ -42,18 +44,74 @@ def build_for(context)
     #
     def item(label, url, options = {})
+      ActiveSupport::Deprecation.warn("Using menu.item in #{@name} context is deprecated. User menu.add_item")
       @items << MenuItem.new(label, url, options)
```
Maybe this should use `add_item` internally?

mrcasals reviewed (approved):
Code looks good to me, but I miss a Changelog entry explaining the changes and how to fix them for developers. I'm mostly blocking the merge because of the Changelog thing.

Thanks for the deprecation, it will help developers for sure 😄 

I left a couple of small comments, please check them out!

mrcasals reviewed (changes requested):
I actually wanted to request those changes, sorry!

alecslupu commented:
> I've reviewed this with @carolromero and we approve this change as @decidim/product 👍🏽
> 
> Just a quick note, I think we should take in account that this is a breaking change as it's changing the API (from `menu.item` to `menu.add_item`). Some alternatives to discuss:
> 
>     1. Add this change to the CHANGELOG so implementers/developers that have modules that use this contracts are aware of this change
>        or
> 
>     2. Change the `menu.add_item` to `menu.item action: :add` and  `menu.remove_item` to `menu.item action: :remove`  where `:add` is the default param (current behaviour)

1. Will be done 
2. I disagree on applying the required change. The remove_item has a total different api than add_item
The `remove_item` only require the idetifier that needs to be removed, whereas the `add_item` has an extra parameter than old `item` parameter. 
At this moment the previous method is still working, but will raise a Deprecation warning.

alecslupu commented:
@mrcasals changes requested have been implemented 
@andreslucena I have added the change log
@oriolgual I have updated the documentation with some information and pictures.

mrcasals commented on decidim-initiatives/lib/decidim/initiatives/admin_engine.rb:
```diff
@@ -104,57 +105,65 @@ class AdminEngine < ::Rails::Engine
+                        active: is_active_link?(decidim_admin_initiatives.initiative_committee_requests_path(current_participatory_space)),
+                        if: current_participatory_space.promoting_committee_enabled? && allowed_to?(:manage_membership, :initiative, initiative: current_participatory_space)
 
           # Maybe we need to add the attachment menu like Participatory process
```
```suggestion
```

mrcasals commented on decidim-elections/lib/decidim/votings/admin_engine.rb:
```diff
@@ -71,63 +72,73 @@ class AdminEngine < ::Rails::Engine
+                        decidim_admin_votings.voting_polling_officers_path(current_participatory_space),
+                        active: is_active_link?(decidim_admin_votings.voting_polling_officers_path(current_participatory_space)),
+                        if: !current_participatory_space.online_voting? && allowed_to?(:read, :polling_officers)
+          menu.add_item :voting_monitoring_committee_members,
```
```suggestion

          menu.add_item :voting_monitoring_committee_members,
```

alecslupu replied:
my understanding is that you requested to add a new line.

mrcasals commented on decidim-elections/lib/decidim/votings/admin_engine.rb:
```diff
@@ -71,63 +72,73 @@ class AdminEngine < ::Rails::Engine
+                        decidim_admin_votings.edit_voting_path(current_participatory_space),
+                        active: is_active_link?(decidim_admin_votings.edit_voting_path(current_participatory_space)),
+                        if: allowed_to?(:update, :voting, voting: current_participatory_space)
+          menu.add_item :edit_voting_landing_page,
```
```suggestion

          menu.add_item :edit_voting_landing_page,
```

alecslupu replied:
My understanding is that you actually requested some spaces here, and i have applied those.

alecslupu replied:
my understanding is that you requested to add a new line.

mrcasals commented on decidim-conferences/lib/decidim/conferences/admin_engine.rb:
```diff
@@ -97,113 +97,131 @@ class AdminEngine < ::Rails::Engine
+          menu.add_item :registrations,
+                        I18n.t("registrations", scope: "decidim.admin.menu.conferences_submenu"),
+                        "#",
+                        active: false, # is_active_link?(decidim_admin_conferences.conference_attachment_collections_path(current_participatory_space)),
```
```suggestion
                        active: false,
```

mrcasals commented on decidim-conferences/lib/decidim/conferences/admin_engine.rb:
```diff
@@ -97,113 +97,131 @@ class AdminEngine < ::Rails::Engine
+                        decidim_admin_conferences.conference_user_roles_path(current_participatory_space),
+                        if: allowed_to?(:read, :conference_user_role, conference: current_participatory_space),
+                        active: is_active_link?(decidim_admin_conferences.conference_user_roles_path(current_participatory_space))
+          menu.add_item :moderations,
```
```suggestion

          menu.add_item :moderations,
```

alecslupu replied:
my understanding is that you requested to add a new line.

mrcasals commented on decidim-conferences/lib/decidim/conferences/admin_engine.rb:
```diff
@@ -97,113 +97,131 @@ class AdminEngine < ::Rails::Engine
+                            allowed_to?(:read, :registration_type, conference: current_participatory_space) ||
+                            allowed_to?(:read, :conference_registration, conference: current_participatory_space),
+                        submenu: { target_menu: :conferences_admin_registrations_menu }
+          menu.add_item :conference_user_roles,
```
```suggestion

          menu.add_item :conference_user_roles,
```

alecslupu replied:
my understanding is that you requested to add a new line.

mrcasals commented on decidim-conferences/lib/decidim/conferences/admin_engine.rb:
```diff
@@ -97,113 +97,131 @@ class AdminEngine < ::Rails::Engine
+                        decidim_admin_conferences.conference_partners_path(current_participatory_space),
+                        if: allowed_to?(:read, :partner, conference: current_participatory_space),
+                        active: is_active_link?(decidim_admin_conferences.conference_partners_path(current_participatory_space))
+          menu.add_item :conference_speakers,
```
```suggestion

          menu.add_item :conference_speakers,
```

alecslupu replied:
my understanding is that you requested to add a new line.

mrcasals commented on decidim-conferences/lib/decidim/conferences/admin_engine.rb:
```diff
@@ -97,113 +97,131 @@ class AdminEngine < ::Rails::Engine
+                        position: 1,
+                        if: allowed_to?(:update, :conference, conference: current_participatory_space),
+                        active: is_active_link?(decidim_admin_conferences.edit_conference_path(current_participatory_space))
+          menu.add_item :components,
```
```suggestion

          menu.add_item :components,
```

alecslupu replied:
my understanding is that you requested to add a new line.

mrcasals commented on decidim-conferences/lib/decidim/conferences/admin_engine.rb:
```diff
@@ -97,113 +97,131 @@ class AdminEngine < ::Rails::Engine
+                        decidim_admin_conferences.conference_conference_registrations_path(current_participatory_space),
+                        active: is_active_link?(decidim_admin_conferences.conference_conference_registrations_path(current_participatory_space)),
+                        if: allowed_to?(:read, :conference_registration, conference: current_participatory_space)
+          menu.add_item :conference_conference_invites,
```
```suggestion

          menu.add_item :conference_conference_invites,
```

mrcasals commented on decidim-conferences/lib/decidim/conferences/admin_engine.rb:
```diff
@@ -97,113 +97,131 @@ class AdminEngine < ::Rails::Engine
+                        decidim_admin_conferences.conference_registration_types_path(current_participatory_space),
+                        active: is_active_link?(decidim_admin_conferences.conference_registration_types_path(current_participatory_space)),
+                        if: allowed_to?(:read, :registration_type, conference: current_participatory_space)
+          menu.add_item :conference_conference_registrations,
```
```suggestion

          menu.add_item :conference_conference_registrations,
```

mrcasals commented on decidim-assemblies/lib/decidim/assemblies/admin_engine.rb:
```diff
@@ -127,85 +130,97 @@ class AdminEngine < ::Rails::Engine
+                        # position: 1,
+                        if: allowed_to?(:read, :space_private_user, assembly: current_participatory_space),
+                        active: is_active_link?(decidim_admin_assemblies.participatory_space_private_users_path(current_participatory_space))
+          menu.add_item :moderations,
```
```suggestion

          menu.add_item :moderations,
```

mrcasals commented on CHANGELOG.md:
```diff
@@ -4,6 +4,8 @@
 
 ### Added
 
+* New Menu Api - [#7545](https://github.com/decidim/decidim/pull/7545)
```
Could you move this to before the `### Added` section and explain what happened? See the changelog for v0.24.0.rc1 for an example:

https://github.com/decidim/decidim/blob/release/0.24-stable/CHANGELOG.md#upgrade-notes

mrcasals reviewed (changes requested):
@alecslupu I left some comments. Most of them are just about improving readability or removing old comments that don't make much sense.

There's also a couple comments on the docs section, and a proposal to improve the Changelog note 😄 

Thanks!!

mrcasals commented on docs/modules/customize/pages/menu.adoc:
```diff
@@ -19,6 +19,316 @@ Depending in the kind of space and module, then you can see or not some of the s
+- sidebar_menu will render the secondary admin section
+- simple_menu will render any submenus
+
+In case you need to identify what menu is being rendered in special page, please
```
It looks like this sentence is unfinished, and there's still the list of all menus down below. Could you complete the sentence and remove the list of menus?

Thanks!

alecslupu replied:
The changes have been implemented, waiting to see if there are any linting issues
